src/winwidget.c

```diff
@@ -673,6 +684,67 @@ void winwidget_destroy(winwidget winwid)
 	return;
 }
 
+#ifdef HAVE_INOTIFY
+void winwidget_inotify_remove(winwidget winwid)
+{
+    if (winwid->inotify_wd >= 0) {
+        D(("Removing inotify watch\n"));
+        if (inotify_rm_watch(opt.inotify_fd, winwid->inotify_wd))
+            eprintf("inotify_rm_watch failed");
+        winwid->inotify_wd = -1;
+    }
+}
+#endif
+
+#ifdef HAVE_INOTIFY
+void winwidget_inotify_add(winwidget winwid, feh_file * file)
+{
+    if (opt.auto_reload) {
+        D(("Adding inotify watch for %s\n", file->filename));
+        char dir[PATH_MAX];
+        feh_file_dirname(dir, file, PATH_MAX);
+        winwid->inotify_wd = inotify_add_watch(opt.inotify_fd, dir, IN_CLOSE_WRITE | IN_MOVED_TO);
+        if (winwid->inotify_wd < 0)
+            eprintf("inotify_add_watch failed");
+    }
+}
+#endif
+
+#ifdef HAVE_INOTIFY
+#define INOTIFY_BUFFER_LEN (1024 * (sizeof (struct inotify_event)) + 16)
+void feh_event_handle_inotify(void)
+{
+    D(("Received inotify events\n"));
+    char buf[INOTIFY_BUFFER_LEN];
+    int i = 0;
+    int len = read (opt.inotify_fd, buf, INOTIFY_BUFFER_LEN);
+    if (len < 0) {
+        if (errno != EINTR)
+            eprintf("inotify event read failed");
+    } else if (!len)
+        eprintf("inotify event read failed");
+    while (i < len) {
+        struct inotify_event *event;
+        event = (struct inotify_event *) &buf[i];
+        for (int j = 0; j < window_num; j++) {
+            if(windows[j]->inotify_wd == event->wd) {
+                if (event->mask & IN_IGNORED) {
+                    D(("inotify watch was implicitely removed\n"));
+                    windows[j]->inotify_wd = -1;
+                } else if (event->mask & (IN_CLOSE_WRITE | IN_MOVED_TO)) {
+                    if (strcmp(event->name, FEH_FILE(windows[j]->file->data)->name) == 0) {
+                        D(("inotify says file changed\n"));
+                        feh_reload_image(windows[j], 0, 1);
```

Comment:
Is there a reason why you're calling feh_reload_image with resize=0, force_new=1?

This way, feh will quit if an image becomes unreadable (e.g. `feh --auto-reload foo.png` and then `echo > foo.png` in another shell). With resize=1, force_new=0, it will instead keep showing the old one and display the new one as soon as a change makes the file valid again. FWIW, this is also the behaviour employed by the regular --reload option, and I think makes sense here as well.

```suggestion
                        feh_reload_image(windows[j], 1, 0);
```